File: SentenceChecker.py
# -*- coding: utf-8 -*-
"""
@author: Patrick DeVries
"""

def checkSentence(wordsArr):
    #Each state is a tuple of (currentState, nextWord, nextState)
    states = [('S0', 'the', 'S1'),
              ('S1', 'lazy', 'S1'),
              ('S1', 'smelly', 'S1'),
              ('S1', 'dog', 'S2'),
              ('S1', 'cat', 'S2'),
              ('S2', 'ate', 'S3'),
              ('S2', 'ran', 'S3'),
              ('S3', 'slowly', 'S4'),
              ('S3', 'noisily', 'S4'),
              ('S3', 'EOI', 'END'),
              ('S4', 'EOI', 'END')]
    
    #auto-fail if the sentence doesnt have minimum of three words ('the', <noun>, <verb>)
    if len(wordsArr) < 3:
        return False
    
    state = 'S0'
    nextW = 0
    
    while(state != 'END'):
        
        #If out of words, return false
        if nextW >= len(wordsArr):
            return False
        
        found = False
        #For all possible states, if the current state of a tuple is this state,
        #Compare the next word to the tuple's next word, and change states if matching
        for i in states:
            if i[0] == state:
                if (wordsArr[nextW].lower() == i[1].lower()):
                    state = i[2]
                    found = True
                    
        #If a state matching the current state and next word was not found, return false
        #Otherwise continue to next word
        if not found:            
            return False
        else:
            nextW = nextW + 1
            continue
        
    #If the function reaches the END state and breaks out of the loop, sentence has passed
    return True


#main function
if __name__=='__main__':  
    #should all be true to pass
    s = [['the', 'lazy', 'dog', 'ate', 'noisily', 'EOI'],
         ['the', 'smelly', 'cat', 'ran', 'slowly', 'EOI'],
         ['the', 'smelly', 'lazy', 'cat', 'ran', 'EOI'],
         ['the', 'smelly', 'lazy', 'smelly', 'dog', 'ate', 'EOI'],
         ['the', 'dog', 'ate', 'EOI'],
         ['the', 'cat', 'ran', 'slowly', 'EOI']]
    
    for i, sentence in enumerate(s):
        # if it is valid, pass
        if (checkSentence(sentence)):
            print('Validation test ' + str(i+1) + ' passed')
        else:
            print('Validation test ' + str(i+1) + ' failed')
            
        print("sentence: ", end = " ")
        for word in sentence:
            print(word, end = " ")
        print('\n')
            
    #should all be false to pass
    s = [['the', 'dog', 'ate'],
         ['dog', 'ate', 'EOI'],
         ['the', 'lazy' , 'ran', 'EOI'],
         ['the'],
         ['EOI'],
         ['the', 'cat', 'ran', 'noisily'],
         ['the', 'dog', 'ate', 'noisily', 'slowly', 'EOI'],
         ['the', 'slowly', 'dog', 'ate', 'noisily', 'EOI'],
         ['the', 'lazy', 'dog', 'ate', 'smelly', 'EOI']]
    
    for i, sentence in enumerate(s):
        #if it is valid, fail
        if (checkSentence(sentence)):
            print('Invalidation test ' + str(i+1) + ' failed')
        else:
            print('Invalidation test ' + str(i+1) + ' passed')
            
        print("sentence: ", end = " ")
        for word in sentence:
            print(word, end = " ")
        print('\n')

    
        
    
    
    

File: SentenceChecker.java

/**
 * Author: Jennifer Carballo
 * Date: April 5, 2020
 * Description: Checks if a string array is a sentence based on sentence structure
 *              Uses HashMap to hold expected states
 * To Run: java -ea SentenceChecker
 */

import java.util.*;
import javafx.util.*;

public class SentenceChecker {

    private static HashMap<Pair<String, String>, String> states;

    public static void main(String[] args) {
        buildStateMap();
        runTests();
    }
    
    /**
     * Puts Pair<Current State, String> 
     * with Next State into HashMap
     * Based off table from previous questions
     */
    public static void buildStateMap() {
        states = new HashMap<>();
        states.put(new Pair("S0", "the"), "S1");
        states.put(new Pair("S1", "lazy"), "S1");
        states.put(new Pair("S1", "smelly"), "S1");
        states.put(new Pair("S1", "dog"), "S2");
        states.put(new Pair("S1", "cat"), "S2");
        states.put(new Pair("S2", "ate"), "S3");
        states.put(new Pair("S2", "ran"), "S3");
        states.put(new Pair("S3", "slowly"), "S4");
        states.put(new Pair("S3", "noisily"), "S4");
        states.put(new Pair("S3", "EOI"), "END");
        states.put(new Pair("S4", "EOI"), "END");
    }

    /**
     * Runs tests on different 
     * correct and incorrect sentences
     * Uses assert
     * Prints "All tests passed!" if tests have...passed obv
     */
    public static void runTests() {
        //True sentences
        String[] sent1 = { "the", "smelly", "dog", "ran", "EOI" };
        String[] sent2 = { "the", "cat", "ran", "EOI" };
        String[] sent3 = { "the", "dog", "ate", "slowly", "EOI" };
        String[] sent4 = { "the", "cat", "ate", "noisily", "EOI" };
        String[] sent5 = { "the", "lazy", "cat", "ate", "slowly", "EOI" };

        //False sentences
        String[] sent6 = { "the", "smelly", "ran", "EOI" };
        String[] sent7 = { "the", "bird", "ran" };
        String[] sent8 = { "dog", "ate", "slowly", "EOI" };
        String[] sent9 = { "the", "cat", "noisily", "EOI" };
        String[] sent10 = { "the", "EOI" };

        try {
            //Should be true
            assert isSentence(sent1);
            assert isSentence(sent2);
            assert isSentence(sent3);
            assert isSentence(sent4);
            assert isSentence(sent5);

            //Should be false
            assert !isSentence(sent6);
            assert !isSentence(sent7);
            assert !isSentence(sent8);
            assert !isSentence(sent9);
            assert !isSentence(sent10);
        } catch (Exception e) {
            System.out.println("Testing Failed. \nInfo:\n" + e.getMessage());
        }
        System.out.println("All tests passed!");
    }

    /**
     * Uses HashMap to compare expected state
     * If all states and string matched and reached "EOI" return true
     * If not, it's not a sentence, return false
     */
    public static Boolean isSentence(String[] sent) {
        Pair<String, String> temp; //Pair<Current State, Word>
        String currState = "S0"; //first state must be S0
        for (int i = 0; i < sent.length; i++) {
            temp = new Pair<>(currState, sent[i]);
            if (states.containsKey(temp)) { //checks if pair is valid
                currState = states.get(temp); //sets next state to current state
                if (currState.equals("END")) { //checks if reached end state
                    return true;
                }
            }
        }
        return false; //sent is not a sentence
    }
}